addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java

```diff
@@ -635,4 +681,260 @@ void shouldFindLocationUrl(String input) {
         // Then
         assertThat(extracted, is(equalTo("http://www.example.com/")));
     }
+
+    /** Unit tests for {@link ExternalRedirectScanRule#extractJsComments(String)}. */
+    @Nested
+    class ExtractJsCommentsUnitTest {
+
+        private static final String BODY_TEMPLATE =
+                """
+                <html><head><script>
+                console.log("something interesting");
+                %s
+                </script></head><body><H1>Redirect</H1></body></html>
+                """;
+
+        private static Stream<Arguments> commentProvider() {
+            return Stream.of(
+                    // Some need to be double escaped because of Java
+                    Arguments.of("Empty line comment", "//", "//"),
+                    Arguments.of("Empty block comment", "/**/", "/**/"),
+                    Arguments.of("Block comment", "/*  comment \\n*/", "/*  comment \\n*/"),
+                    Arguments.of(
+                            "Line comment with CRLF",
+                            "console.log('x'); // comment\\r\\nconsole.log('y');",
+                            "// comment\\r\\nconsole.log('y');"),
+                    Arguments.of(
+                            "Block comment containing line terminator + line comment",
+                            "/* block start\\n// inside block */ console.log('x');",
+                            "/* block start\\n// inside block */"),
+                    Arguments.of(
+                            "Escaped quote before comment",
+                            "console.log('it\\'s fine'); // real comment",
+                            "// real comment"),
+                    Arguments.of(
+                            "Escaped backslash before comment",
+                            "console.log('c:\\\\'); // comment",
+                            "// comment"),
+                    Arguments.of("Single line", "// comment ", "// comment"),
+                    Arguments.of(
+                            "Block inside Single line", "// /* comment; */", "// /* comment; */"),
+                    Arguments.of(
+                            "Single line inside Block comment",
+                            "/*  comment \\n // example */",
+                            "/*  comment \\n // example */"),
+                    Arguments.of(
+                            "Inline block",
+                            "console.log(\"example\"); /* console.log('comment'); */",
+                            "/* console.log('comment'); */"),
+                    Arguments.of(
+                            "Inline incomplete block",
+                            // What do we really want in this case?
+                            "console.log(\"example\"); /* console.log('comment'); ",
+                            "/* console.log('comment'); \n</script></head><body><H1>Redirect</H1></body></html>"),
```

Comment:
For this rule I think we only deal with JS inside HTML, so capturing to the closing script tag is probably reasonable.

In the future we could have a boolean that indicates if the end of script tag is used, or some other behavior to pure JS files.